### ISSUE TITLE
Adjoint solver with nonreciprocal material

### ISSUE THREAD

**Issue by Edith-Z** [NanoComp/meep#1548]
Is MEEP available for adjoint methed with nonreciprocal material? And can I directly add a material use the anisotropic complex epsilon tensor? I use Ce:YIG and the permittivity is:
[[4.8841+0.j         0.    +0.j         0.    +0.00761222j]
 [0.    +0.j         4.8841+0.j         0.    +0.j        ]
 [0.    -0.00761222j 0.    +0.j         4.8841+0.j        ]]

**Comment by Edith-Z** [comment]
I only need to simulate at 1.55μm. Does this work?

import meep as mp
import math
n_ceyig = 2.21
theta = -400000
lbd = 1550e-9
gy = -lbd*n_ceyig*theta/180
epsn = 2.21**2
f0 = 1/1.55
fn   = 1    # natural frequency
omg0 = f0*2*math.pi
omgn = fn*2*math.pi
gamma = 0   # damping rate
by = 5 
sigma = gy*((omgn**2-omg0**2)**2-omg0**2*by**2)/omgn**2/omg0/by
eps1=epsn-omgn**2*(omgn**2-omg0**2)*sigma/((omgn**2-omg0**2)**2-omg0**2*by**2)
eps3=eps1
eps2=epsn-omgn**2*sigma/(omgn**2-omg0**2)
susc = [mp.GyrotropicLorentzianSusceptibility(frequency=fn, gamma=gamma, sigma=sigma,
                                              bias=mp.Vector3(0, by, 0))]
ce_yig = mp.Medium(epsilon_diag=mp.Vector3(eps1, eps2, eps3),E_susceptibilities=susc,mu=1)

**Comment by stevengj** [comment]
> And can I directly add a material use the anisotropic complex epsilon tensor?

No, because you have to have dispersion to get a passive complex material in the time domain.

> Is MEEP available for adjoint methed with nonreciprocal material?

In principle, this should work, but I don't think it is implemented yet.   In principle, for the adjoint calculation you need to "transpose" the nonreciprocal material (e.g. flip the sign of the **b** vector in a Gyrotropic Drude-Lorentz model).

This would be easy to implement, but I don't think it has been done yet, right @smartalecH?

**Comment by smartalecH** [comment]
> In principle, for the adjoint calculation you need to "transpose" the nonreciprocal material

I think what you are saying is that the material properties need to be transposed in the adjoint *run*, right? (adjoint calculation is a little ambiguous to me, since some people refer to the recombination step when they say "calculation"). 

We don't do anything like that right now, unfortunately. But it might be nice to have a callback of sorts that manually transposes these materials.

In the recombination step (λᵀ ∂A/∂ρ x) where the material effects typically come into play (e.g. dispersion), we currently calculate ∂A/∂ρ (needed in the recombination step) using a finite difference:

https://github.com/NanoComp/meep/blob/bcb90f30951c1a4a2d008ced0cd32e159b3e0098/src/meepgeom.cpp#L2589-L2615

So it's generalizable to any type of material in that sense.

**Comment by stevengj** [comment]
Yes, they need to be transposed in the adjoint simulation itself, not in the subsequent use of the adjoint solution to find the gradient.
